documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,823 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+`SupportedFlags` is a bitmask containing other supported options for the cursor which can be mixed and matched if supported. This must be queried before setting `Flags` - the currently active options. An exception should be thrown if an attempt is made to set an unsupported flag on `Flags`. Unlike the other properties, `Flags` can have multiple bits set.
+
+`Flags` replaces `IsConfined`, which was undefined behaviour on platforms where it was not supported in 2.X - now with `SupportedFlags` you can query its support before setting it.
+
+### Enums
+
+```cs
+[Flags]
+public enum CursorModes
+{
+    Normal,
+    Hidden = 1 << 0,
+    Disabled = 1 << 1,
+    Raw = 1 << 2
+}
+```
+```cs
+[Flags]
+public enum CursorStyles
+{
+    Default,
+    Arrow = 1 << 0,
+    IBeam = 1 << 1,
+    Crosshair = 1 << 2,
+    Hand = 1 << 3,
+    HResize = 1 << 4,
+    VResize = 1 << 5,
+    Custom = 1 << 6,
+}
+```
+```cs
+[Flags]
+public enum CursorFlags
+{
+    None,
+    Confined = 1 << 0
+}
+```
+
+# Keyboard Input
+
+Once again, the interface is very simple.
+
+```cs
+public interface IKeyboard : IInputDevice
+{
+    ref readonly KeyboardState State { get; }
+    string? ClipboardText { get; set; }
+    void BeginInput();
+    void EndInput();
+}
+```
+
+`State` is the device state as defined earlier.
+
+`BeginInput` starts recording textual input, bringing up the on-screen keyboard on platforms where this is needed (i.e. mobile). The `State.Text` property will not be populated until after this is called.
+
+`ClipboardText` allows getting and setting the text on the user's clipboard so they can paste information to/from your application in others.
+
+`KeyboardState` is defined as follows:
+
+```cs
+public readonly record struct KeyboardState
+(
+    InputReadOnlyList<char>? Text,
+    KeyState Keys
+);
+```
+
+`Text` contains the characters typed on the keyboard since `IKeyboard.BeginInput`, and accounts for backspaces. This is cleared (set to `null`) when `IKeyboard.EndInput` is called, and will not be non-`null` again until another `IKeyboard.BeginInput` call. Given that `KeyChar` events are raised one character at a time, this property will update one character at a time to keep the state consistent with the event.
+
+**INFORMATIVE TEXT:** This is something we can optimize in `InputList` to not be allocatey, rest assured it is not acceptable to the Silk.NET team to allocate a new list for every character.
+
+```cs
+public readonly record struct KeyState
+(
+    InputReadOnlyList<Key> Down
+)
+{
+    public bool this[KeyName btn] { get; }
+    public bool this[int scancode] { get; }
+    public InputReadOnlyList<Key> Up { get; }
```

Comment:
I'd get rid of `KeyState.Up`. It would make more sense if we had gone purely state-based, but having the events now this just seems like bloat

Reply:
> `Up` and the indexer will be implemented in terms of `Down`, which is the only property that a backend will need to set. Scancode-only keys will not be present in `Up`.

Also, why would it not include scancode-only keys? the `Key` struct contains both `KeyName` and scancode

Reply:
On `KeyState.Up`, we can get rid of this but this is mostly for parity with the other button states. This implementation of this won't be too hard or inefficient, either. There is probably a case where a user would want to use `KeyState.Up` that I haven't thought of, admittedly it's far less likely than the other states but just wanted to make sure all the states looked the same just in case.

Reply:
No, that's because we'll have to have `InputReadOnlyList` contain one `Key` for every single value of `int` i.e. every single scancode.

Reply:
Hmmm... actually it is possible we could implement this in an efficient way but I still feel it only makes sense to include named keys.

Reply:
Wait, is `KeyState.Up` supposed to be "the keys that were `Down` last update but aren't this update" or is it supposed to be "The keys that are not in `Down`"?

Reply:
"The keys that are not in `Down`"

Reply:
How would we efficiently implement this? Remember it has to implement `IReadOnlyList`, so it has to be enumerable, contain a `Count`, and access by index.

Reply:
**INFORMATIVE TEXT:**
- for index:
    1. assume that the index is the final scancode.
    2. iterate over `Down`.
        a. if the scancode is smaller than the index, decrement the index by one
    3. return a `Key` containing the index as the scancode and a named mapping if possible
- for enumerator:
    - either iterate over all keynames or if we do decide to include unnamed keys iterate over all values of `int`
...

now that i'm typing this i realise this is entirely stupid. we should just only have `Down` everywhere. Agree?

Reply:
Haha, agreed

Reply:
Removed now.